From dbbcc1503abd239a22ba49afd39bc7f5584ce4dc Mon Sep 17 00:00:00 2001
From: Pascal Delameillieure <pascal@kubrick.be>
Date: Thu, 2 May 2024 11:47:57 +0200
Subject: [PATCH] Make lazy loading more like the native browser behavior

---
 js/resizer.js | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/js/resizer.js b/js/resizer.js
index c91fed5..27a2736 100644
--- a/js/resizer.js
+++ b/js/resizer.js
@@ -17,7 +17,13 @@

       // Define an offset to preload lazy loaded images before they are visible
       // in the viewport.
-      const preloadOffset = 300;
+      // Offsets are based on connection as documented on
+      // https://web.dev/articles/browser-level-image-lazy-loading#distance-from-viewport
+      const connection = navigator.connection;
+      const connectionType = connection?.effectiveType;
+      const connectionSaveData = connection?.saveData;
+
+      const preloadOffset = connectionType === "3G" || connectionSaveData ? 2500 : 1250;

       // Find the best suitable image to display.
       const updateImage = function (image) {
--
GitLab
